### PR TITLE
Document dev checklist

### LINKED ISSUE #212: Application architecture
The way the codebase is currently structured makes sense for a small, short lived project, but will become increasingly unsustainable as the project becomes larger and requires more features to be added. 

I'd suggest us _very gradually_ moving closer to something like the following outlines:

### Root
```
./
|- docs         Documentation
|- client       Client app
|- config       Project configuration
|- migrations   Database migration files
|- scripts      Developer scripts (like building/destroying db)
|- server       Server app
|- tests        ...Tests
|- ...dotfiles
|- package.json Single top-level package.json instead of two different ones
|- README.md
```

### Client
```
./client
|- assets             Raw assets (img, external or pre-compiled js/css, etc.)
|- dist               Built assets will be stored and served from here
|- css                Styling source files
|- src
|  |- admin           All code specific to the admin UI
|  |  |- components
|  |  |- actions
|  |  |- ...
|  |  |- index.js     Exports a single component that contains the entire admin UI
|  |- shared          All code shared between the admin and visitor UIs
|  |  |- components
|  |  |- actions
|  |  |- ...
|  |- visitor         All code specific to the visitor UI
|  |  |- components
|  |  |- actions
|  |  |- ...
|  |  |- index.js     Exports a single component that contains the entire visitor UI
|  |- App.js
|  |- index.js
```

### Server
```
./server
|- admin               All code related to admin features
|  |- controllers      Route definitions
|  |  |- api           Route definitions for API routes
|  |- models           Database models (objects representing admin-related entities in the DB)
|  |- index.js         Exports single function/app that can be mounted onto the main express app
|- community_business  All code related to CB features
|  |- controllers
|  |- ...
|- visitor             All code related to visitor features
|  |- controllers
|  |- ...
|- shared              All shared code
|  |- db
|  |- models
|  |- ...
|- index.js
```

The idea is to make each feature self contained as far as possible, and make the core of the app kind of like a generic hub onto which new features can be bolted with minimal faff.

This also helps with maintenance. For example, if I'm working on changing some existing feature for the admin, I want to be able to:
1. Find (_all of_) the relevant code quickly
2. Be confident in determining which code is used solely by admin-features and which is shared or may have side-effects

@RogeredBacon @astroash Keen to hear both your thoughts, given you've worked on this longer than me. Give me criticism, tell me if I've missed something, or if any of the above has broader implications that I might not be aware of (other than being a lot of work :wink:)

Can provide diagram if you like.

### PR DESCRIPTION
**Pull request by eliascodes**
Fixes #212 

* Documents the developer checklist as discussed
* Documents the desired project architecture

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+|- package.json Single top-level package.json instead of two different ones
+|- README.md
+```
+
+### Client
+```
+./client
+|- assets             Raw assets (img, external or pre-compiled js/css, etc.)
+|- dist               Built assets will be stored and served from here
+|- css                Styling source files
```
Comment (astroash) [inline docs/architecture.md]: is this dependant on whether we use styled elements?

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+|- src
+|  |- admin           All code specific to the admin UI
```
Comment (astroash) [inline docs/architecture.md]: can we specify that this is cb admin?

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+|  |  |- ...
+|  |  |- index.js     Exports a single component that contains the entire visitor UI
+|  |- App.js
+|  |- index.js
+```
+
+### Server
+```
+./server
+|- admin               All code related to admin features
```
Comment (astroash) [inline docs/architecture.md]: admin -> cb_admin?

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+|  |- controllers      Route definitions
+|  |  |- api           Route definitions for API routes
+|  |- models           Database models (objects representing admin-related entities in the DB)
+|  |- index.js         Exports single function/app that can be mounted onto the main express app
+|- community_business  All code related to CB features
```
Comment (astroash) [inline docs/architecture.md]: Or is the top one for twine_admin and this one for cb_admin?

OR... are you separating out the login needed for cb's standard login and their admin section...

File: docs/dev_checklist.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Development Checklist
+
+This is a checklist of things to think about before considering your piece of work "done".
```
Comment (astroash) [inline docs/dev_checklist.md]: "done" -> _**"done"**_ 😉

File: docs/dev_checklist.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+## Adding New Features
+* [ ] Is my new feature sufficiently well tested?
+* [ ] Have I updated the project documentation and README to reflect my changes?
+* [ ] Does the new feature conform to the desired project architecure?
```
Comment (astroash) [inline docs/dev_checklist.md]: ewwww, "conform"...

File: docs/architecture.md (hunk 1 above)
Comment (eliascodes) [reply docs/architecture.md]: Yes, if we use styled elements, this will probably disappear.

File: docs/architecture.md (hunk 2 above)
Comment (eliascodes) [reply docs/architecture.md]: I think this was meant to be twine-admin. Will update.

File: docs/architecture.md (hunk 4 above)
Comment (eliascodes) [reply docs/architecture.md]: The former

File: docs/dev_checklist.md (hunk 6 above)
Comment (eliascodes) [reply docs/dev_checklist.md]: unconstructive...

Comment (astroash) [reply docs/dev_checklist.md]: 🙄

Comment (eliascodes) [reply docs/dev_checklist.md]: What word would you like me to use?

**Pull request merged**